File: anagram.erl
#!/usr/bin/env escript


main([File, LoadAttempts, FindAttempts]) -> main_func(File, list_to_integer(LoadAttempts), list_to_integer(FindAttempts));
main(_)      -> main_func("words", 15, 100).

main_func(File, LoadAttempts, FindAttempts) ->
    ets:new(words_ets, [set, named_table, public]),
    LoadAverage = measure_average(fun()-> load_file(File) end, LoadAttempts),
    %io:format("-> ~p ~n", [ets:tab2list(words_ets)]),
    io:format("~naverage load time: ~p (mks), ~p load attempts ~n", [LoadAverage, LoadAttempts]),
    finder(FindAttempts).

load_file(FileName) ->
    ets:delete_all_objects(words_ets),
    case file:open(FileName, [read]) of
        {ok, File} -> read_lines(File)
        ;_         -> io:format("cannot open file ~p ~n", [FileName]), exit(cannot_open_file)
    end.

read_lines(File) ->
    case io:get_line(File, "") of
        eof  -> io:format(".", []);
        Line -> save_word(trim(Line)), read_lines(File)
    end.

save_word(Word) ->
    LWord = lists:sort(string:to_lower(Word)),
    case ets:lookup(words_ets, LWord) of
        [{LWord, List}] -> ets:insert(words_ets, {LWord, [Word]++List})
        ;_              -> ets:insert(words_ets, {LWord, [Word]})
    end.

finder(FindAttempts) ->
    case io:get_line("Enter word:") of
        "quit\n" -> io:format("Bye ~n", []);
        Word     ->
            W = string:to_lower(trim(Word)),
            %io:format("Candidates: ~p~n", [lists:usort(c(W))]),
            FindAverage = measure_average(fun() -> find_anagram(W) end, FindAttempts),
            case find_anagram(W) of
                [] -> io:format("not found~n", []);
                L  -> io:format("anagrams found: ~p ~n", [string:join(L, ", ")])
            end,
            io:format("find average: ~p (mks), ~p find attempts ~n", [FindAverage, FindAttempts]),
            finder(FindAttempts)
    end.

find_anagram(W) ->
    SW = lists:sort(W),
    case ets:lookup(words_ets, SW) of
        [{_, List}] -> [X || X<- List, string:to_lower(X) =/= W]
        ;_          -> []
    end.

trim(String) -> re:replace(String, "\\s+", "", [global, {return, list}]).

measure_average(Fun, Times) -> measure_average(Fun, Times, undefined).
measure_average(_, 0, A) -> A;
measure_average(Fun, Times, Average) ->
    T1 = now(), Fun(), T2 = now(),
    Ret = timer:now_diff(T2, T1),
    Average1 = if is_integer(Average) -> (Average + Ret) div 2; true -> Ret end,
    measure_average(Fun, Times - 1, Average1).
    

File: anagram.pl
#!/usr/bin/perl

use strict;
use warnings;
use Data::Dumper;
use Time::HiRes qw(gettimeofday tv_interval);

my %words = ();
my $filename = $ARGV[0] || "words";
my $load_file_attempts = $ARGV[1] || 15;
my $find_attemts = $ARGV[2] || 100;

my ($load_average_time, $find_average_time);

for(1..$load_file_attempts){
    my $start_time = [gettimeofday()];
    loadfile($filename);
    my $diff_mks = int(tv_interval($start_time) * 1000000);
    $load_average_time = $load_average_time?int(($load_average_time+$diff_mks)/2):$diff_mks;
    print "."
}
print "\nLoad average time ${load_average_time} mks, ${find_attemts} load attemts \n";

while(1){
    print "enter word:\n";
    my $line = <STDIN>;
    chomp($line);
    last if $line eq 'quit';
    my $ret;
    for(1..$find_attemts){
        my $start_time = [gettimeofday()];
        $ret = finder($line);
        my $diff_mks = int(tv_interval($start_time) * 1000000);
        $find_average_time = $find_average_time?int(($find_average_time+$diff_mks)/2):$diff_mks;
    }
    print $ret."\n";
    print "Average find time: ${find_average_time} mks, ${find_attemts} find attempts \n";
    undef($find_average_time)
}

sub loadfile{
    %words = ();
    open(DATA, shift) or die "cannot open file";
    while(<DATA>){chomp; push @{$words{join("",sort(split(//,lc($_))))}},$_}
    close(DATA);
}

sub finder{
    my $line = shift;
    my $key = join("",sort(split(//,lc($line))));
    my $anagrams = $words{$key};
    $anagrams?"Found anagrams: ".join(", ", grep {lc($_) ne lc($line)} @$anagrams):"not found";
}
